### PR TITLE
Make RemoteStoreReplicationSource#getSegmentFiles asynchronous

### PR DESCRIPTION
**Pull request by mch2**
### Description
This change makes Make RemoteStoreReplicationSource#getSegmentFiles asynchronous.
It also ensures that SegmentReplicationTargets are properly closed during cancellation with remote store enabled if the cancel is issued during a file download.

### Related Issues
Resolves https://github.com/opensearch-project/OpenSearch/issues/8089
Resolves https://github.com/opensearch-project/OpenSearch/issues/10619

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by mch2** [comment]
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** FAILURE ❌
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/28486/
> * **CommitID:** [cd06a0d](https://github.com/opensearch-project/OpenSearch/commit/cd06a0d78faadba00860c3f7327f5ee0a2d4b18e)
>   Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green.
>   Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

https://github.com/opensearch-project/OpenSearch/issues/6090

**Review by andrross — COMMENTED** [review]
This looks good to me! But I authored one of the commits in this PR so I'd love to have approval from another maintainer.

File: server/src/main/java/org/opensearch/indices/replication/common/ReplicationTarget.java
Comment (kotwanikunal) [inline server/src/main/java/org/opensearch/indices/replication/common/ReplicationTarget.java:288]: Can this leave the store open and close the remote store in a non-happy case?

**Review by kotwanikunal — APPROVED** [review]

File: server/src/main/java/org/opensearch/indices/replication/common/ReplicationTarget.java
Comment (kotwanikunal) [reply server/src/main/java/org/opensearch/indices/replication/common/ReplicationTarget.java:288]: Considering the dependency and linking of remote store - store - should it rather be 
```
 try {
            store.decRef();
            if (indexShard.indexSettings().isRemoteStoreEnabled()) {
                indexShard.remoteStore().decRef();
            }
       }
```

Comment (mch2) [reply server/src/main/java/org/opensearch/indices/replication/common/ReplicationTarget.java:288]: if the store's refcount hits zero it will call closeInternal which throws if it can't close the directory:

```java
    private void closeInternal() {
        // Leverage try-with-resources to close the shard lock for us
        try (Closeable c = shardLock) {
            try {
                directory.innerClose(); // this closes the distributorDirectory as well
            } finally {
                onClose.accept(shardLock);
            }
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }
```
If that were to throw we'd still need to decref the remoteStore.

**Pull request merged**
